tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc

```diff
@@ -120,6 +120,135 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+  //
+  std::vector<HloInstruction*> users = concat->users();
+  bool all_1d_slices = absl::c_all_of(users, [](const HloInstruction* i) {
+    return Is1dSliceWithoutStrides(i);
+  });
+  if (!all_1d_slices) {
+    // Limit our supported cases to 1 dimensional slices.
+    return false;
+  }
+  absl::c_sort(users, [](const HloInstruction* a, const HloInstruction* b) {
+    return a->slice_starts().at(0) < b->slice_starts().at(0);
+  });
+
+  // Verify that each operand to the concat is reversed by a slice.
+  if (users.size() != concat->operand_count()) {
+    return false;
+  }
+  int64 prev_limit = 0;
+  for (size_t i = 0; i < users.size(); ++i) {
+    const HloInstruction* u = users[i];
+    int64 slice_size = u->slice_limits().at(0) - u->slice_starts().at(0);
+    if (u->slice_starts().at(0) != prev_limit ||
+        slice_size != ShapeUtil::ElementsIn(concat->operand(i)->shape())) {
+      return false;
+    }
+    prev_limit = u->slice_limits().at(0);
+  }
+
+  return true;
+}
+
+// Returns whether we can prove the transitive uses are in effect elementwise
+// operations. A concat followed by slices is considered (effectively)
+// elementwise if the slices combiningly is a reverse function of the concat.
+// We can prove more patterns but we currently do just enough for
+// SliceInputFusion.
+bool AreTransitiveUsesEffectivelyElementwise(const HloInstruction* instr) {
+  absl::flat_hash_set<const HloInstruction*> visited;
+  absl::InlinedVector<const HloInstruction*, 4> stack;
+  stack.push_back(instr);
+  while (!stack.empty()) {
+    const HloInstruction* current = stack.back();
+    stack.pop_back();
+    visited.insert(current);
+    for (const HloInstruction* user : current->users()) {
+      VLOG(3) << "Visiting: " << user->ToString();
+      switch (user->opcode()) {
+        case HloOpcode::kTuple:
+        // We say reshape is fine because it does not reorder elements.
+        case HloOpcode::kReshape:
+          break;
+        case HloOpcode::kConcatenate:
+          if (!ConcatHasNoEffect(user)) {
+            return false;
+          }
+          break;
+        case HloOpcode::kSlice:
+          if (user->operand(0)->opcode() != HloOpcode::kConcatenate) {
+            return false;
+          }
+          break;
+        default:
+          if (user->IsElementwise()) {
+            for (const int64 use_index : user->OperandIndices(current)) {
+              if (!user->IsElementwiseOnOperand(use_index)) {
+                // Found a user that is non-elementwise on the current
+                // instruction.
+                return false;
+              }
+            }
+          } else {
+            VLOG(3) << "Cannot prove that the op is effectively elementwise: "
+                    << user->ToString();
+            return false;
+          }
+          break;
+      }  // end of switch
+      if (user->opcode() != HloOpcode::kTuple &&
+          !LayoutUtil::IsMonotonicWithDim0Major(user->shape().layout())) {
```

Comment:
This is a more restricted form. So long as all layouts are row major, we ensure no layout change. The HorizontalLoopFusion only supports row-major layouts. However, since this creates confusion, I will change to what you suggest.